junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt

```diff
@@ -95,8 +95,11 @@ fun assertAll(heading: String?, vararg executables: () -> Unit) =
  * ```
  * @see Assertions.assertThrows
  */
-inline fun <reified T : Throwable> assertThrows(noinline executable: () -> Unit): T =
-    Assertions.assertThrows(T::class.java, Executable(executable))
+inline fun <reified T : Throwable> assertThrows(executable: () -> Unit): T {
+    val result = runCatching(executable)
+
+    return Assertions.assertThrows(T::class.java) { result.getOrThrow() }
+}
```

Comment:
Sorry, just had this realization. `runCatching` was introduced in Kotlin 1.3.

Does this have the implication that this method will now only be supported in Kotlin 1.3 and higher?

Reply:
This shouldn't be a problem, since coroutines become stable in [1.3](https://blog.jetbrains.com/kotlin/2018/10/kotlin-1-3/), and this method designed for coroutines users

Reply:
Hmm, good question. We have this integration test in place that's [compiled against Kotlin 1.1](https://github.com/junit-team/junit5/blob/b01802956c05822d59ec73cbf377529a3abfb5f8/platform-tooling-support-tests/projects/gradle-kotlin-extensions/build.gradle.kts#L27-L28). Shouldn't that be broken? 🤔 
https://github.com/junit-team/junit5/blob/3f7fed61f2edcee2a5551dbde75382b1dbe21267/platform-tooling-support-tests/projects/gradle-kotlin-extensions/src/test/kotlin/com/example/project/ExtensionFunctionsTests.kt

Reply:
Or does that work because we use the 1.3.61 version of stdlib?
https://github.com/junit-team/junit5/blob/b01802956c05822d59ec73cbf377529a3abfb5f8/platform-tooling-support-tests/projects/gradle-kotlin-extensions/build.gradle.kts#L20

Reply:
We could probably adjust the implementation to use the same behaviour as `runCatching` without it, though it's a bit more verbose that I had hoped.

```kotlin
inline fun <reified T : Throwable> assertThrows(executable: () -> Unit): T {
  val throwable: Throwable? = try {
    executable()
  } catch (caught: Throwable) {
    caught
  } as? Throwable

  return Assertions.assertThrows(T::class.java) {
    if (throwable != null) {
      throw throwable
    }
  }
}
```

Reply:
@ashdavies Could you try to make the implementation test fail?

Reply:
Resolves in `master` in c5d658c5aa8e5550410e9691d90e3216fe9a55e4.

Reply:
Thanks for addressing, hadn't yet had the chance to check

Reply:
No worries!